tests/components/auth/test_indieauth.py

```diff
@@ -83,8 +83,6 @@ def test_client_id_hostname() -> None:
     assert indieauth._parse_client_id("http://192.168.0.0")
     assert indieauth._parse_client_id("http://192.168.255.255")
 
-    with pytest.raises(ValueError):
-        assert indieauth._parse_client_id("http://255.255.255.255/")
```

Comment:
This is a semantics change.  Maybe we should keep a check for this

Reply:
Is it really? 255.255.255.255 is "limited broadcast", and has been reserved from being used for unicast traffic for the past... ~40 years (RFC 919 section 7 dated October 1984). I'm not even sure how usable it'll actually be even if you force your OS to use it, but if someone really managed to, I guess they get to keep the pieces when things break? (As a sidenote, we are already (purposefully) diverging from the indieauth spec, which prohibits bare IP addresses entirely, if this wasn't an edge case enough.)

Reply:
I'm assuming `255.255.255.255` was chosen for a reason in that test so it seems wrong to remove it without letting the original author of the PR that it was added in have chance to review (#15369). Alternatively, we can keep compatibility with something like:

```diff
diff --git a/homeassistant/util/network.py b/homeassistant/util/network.py
index d5830d25b69..3360347b1b4 100644
--- a/homeassistant/util/network.py
+++ b/homeassistant/util/network.py
@@ -7,6 +7,8 @@ import re
 
 import yarl
 
+BROADCAST = ip_address("255.255.255.255")
+
 
 def is_loopback(address: IPv4Address | IPv6Address) -> bool:
     """Check if an address is a loopback address."""
@@ -16,7 +18,12 @@ def is_loopback(address: IPv4Address | IPv6Address) -> bool:
 
 def is_private(address: IPv4Address | IPv6Address) -> bool:
     """Check if an address is a unique local non-loopback address."""
-    return address.is_private and not is_loopback(address) and not address.is_link_local
+    return (
+        address.is_private
+        and not is_loopback(address)
+        and not address.is_link_local
+        and address != BROADCAST
+    )
 
 
 def is_link_local(address: IPv4Address | IPv6Address) -> bool:
```

Reply:
I strongly believe we should not be adding those kind of "special" semantics, especially for reasons that are not clear. 255.255.255.255 is *not* routable, per RFC 919, section 7, and RFC8190. You literally cannot use it, e.g. try a traceroute from your computer, it will refuse to even route the traffic to your default gateway.

The fact that HA's `is_private` returned `False` before, was a bug. Python's `is_private` is following those RFCs (and IANA's special-purpose registry assignments) and has the correct semantic here.

I can't imagine a scenario where bug-for-bug compatibity is desired, but I acknowledge I may be missing something! @baloob I've noticed you're the #15369/indieauth author, perhaps you could shed some light on why http://255.255.255.255/ was chosen as a test case?

BTW, the only case in this entire file where I see the need to override stdlib, is with `is_loopback`, and that is due to a Python stdlib bug.  But in this case, I've sent a PR to Python to fix that in stdlib itself and over time remove our custom `is_loopback` override. I intend to entirely deprecate the rest of these utility functions in a subsequent PR.

Reply:
I suggest keeping bug-for-bug compatibility for now as it makes this PR mergable now.  It can be removed in a followup when others are available to review.

Reply:
Thanks for the quick response & review! Honestly I don't think that makes sense. This whole PR is about resolving long-standing bugs resulting from HA hardcoding IANA-reserved ranges, and not doing it right (either in the first place, or due to drift). Reintroducing some of these bugs without a clear rationale is a slippery slope and kind of defeats the whole point. I have tried to keep semantics in a couple of places where it made sense (see the commit message), but in the 255.255.255.255 example in particular, I don't think it does.

As demonstrated above, 255.2555.255.255 is unroutable in Linux, including in HA's Docker/HassOS, so there is no realistic way one can use this IP in a network, and still exchange traffic with HA. IOW, the test suite tests for something that is just not possible. I hope this demonstration makes the change easier to review, but if you'd like to wait for additional reviewers, I understand :)

Reply:
Probably best to wait for another reviewer as I don't feel comfortable making that decision in isolation.

Reply:
I kinda agree with bdraco, and judging by your comments, you do too:

> Is it really? 255.255.255.255 is "limited broadcast", and has been reserved from being used for unicast traffic for the past... ~40 years (RFC 919 section 7 dated October 1984).

Meaning this should have raised on this given test. This test should be re-instated, and the code should be fixed to raise for this case.

Reply:
Thanks for the review! Ι'll admit that I'm not super familiar with indieauth, and I may be misunderstanding something. And I'd like to add code that I understand :)

First of all: the indieauth code accepts only "local" addresses for auth. The networking code previously (erroneously) considered this address "globally routable" and thus indieauth did not accept it. The address is definitely not globally routable. Is this all correct?

Second: should 255.255.255.255 be singled out in the indieauth code? Why? What's the real-life 
condition under which this address would be used? Basically, under which scenario would this code path be reached (outside of tests)?